core/src/replay_stage.rs

```diff
@@ -974,31 +982,38 @@ impl ReplayStage {
     // Checks for and handle forks with duplicate slots.
     fn process_duplicate_slots(
         duplicate_slots_receiver: &DuplicateSlotReceiver,
+        duplicate_slots_tracker: &mut DuplicateSlotsTracker,
         gossip_duplicate_confirmed_slots: &GossipDuplicateConfirmedSlots,
         bank_forks: &RwLock<BankForks>,
         ancestors: &HashMap<Slot, HashSet<Slot>>,
         descendants: &HashMap<Slot, HashSet<Slot>>,
         progress: &mut ProgressMap,
         fork_choice: &mut HeaviestSubtreeForkChoice,
     ) {
-        let duplicate_slots: Vec<Slot> = duplicate_slots_receiver.try_iter().collect();
+        let new_duplicate_slots: Vec<Slot> = duplicate_slots_receiver
+            .try_iter()
+            .filter(|duplicate_slot|
+                // If this is the first time seeing this duplicate, run the processing
+                !duplicate_slots_tracker.insert(*duplicate_slot))
```

Comment:
This filtering seems unnecessary since duplicate slots are sent if they were already detected in blockstore, right? Also, I think it would be cleaner to do the duplicate_slot insert inside the `check_slot_agrees_with_cluster` method (before the bank hash check)

Reply:
Updated!